### PR TITLE
video_layout (mame artwork)

### PR DESCRIPTION
**Pull request by huwpascoe**
## Description
* adds video_layout module
* now has menu settings

## Related Issues
https://github.com/libretro/RetroArch/issues/7419

### CODE REVIEW

**Comment by twinaphex** [comment]
OK, so does this already work ? Are there some examples of it working?

**Comment by huwpascoe** [comment]
@twinaphex

Latest status
https://github.com/libretro/RetroArch/issues/7419#issuecomment-478306671

**Comment by huwpascoe** [comment]
Finished. Ready for review/merge.

**Comment by grant2258** [comment]
is working well testes invaders with fba and plus no isssues. Only problem is you will need to disable any shaders this is to be expected considering the image is rotated and scaled then scaled again

**Comment by bparker06** [comment]
not C89 compliant

File: gfx/video_driver.c [hunk 1]
```diff
@@ -1099,6 +1107,16 @@ static bool video_driver_init_internal(bool *video_is_threaded)
    command_event(CMD_EVENT_OVERLAY_DEINIT, NULL);
    command_event(CMD_EVENT_OVERLAY_INIT, NULL);
 
+#ifdef HAVE_VIDEO_LAYOUT
+   video_layout_driver_init(video_driver_data, current_video->video_layout_render_interface(video_driver_data));
```
Comment (bparker06) [inline gfx/video_driver.c]: why isn't this inside the if statement? we don't want to allocate memory we'll never use

**Comment by huwpascoe** [comment]
> Not C89 compliant.
> 
> Please make sure to thoroughly test this using ASAN on Linux as well.

I've been using gflags on windows if that's similar?

File: gfx/video_driver.c (hunk 1 above)
Comment (huwpascoe) [reply gfx/video_driver.c]: Reason is the interface is set at that point, if it were disabled on startup then enabled in menu, it'd crash. I'll see if there's a public interface to get video_driver_data and current_video

Comment (huwpascoe) [reply gfx/video_driver.c]: Exposed the interface so now enable/disable connects directly to init/deinit

**Comment by huwpascoe** [comment]
Okay looks like the C89 builds have succeeded. (the other broken builds are unrelated to this PR I think)

**Comment by huwpascoe** [comment]
fixed pedantic errors, rebased.

**Comment by huwpascoe** [comment]
rebased

**Comment by markwkidd** [comment]
@twinaphex I can see that @huwpascoe will need to resolve a merge conflict. can you say whether there are any other changes needed for this PR to be merged?

**Comment by hizzlekizzle** [comment]
I spoke with Twinaphex about this on discord recently. He's wanting to move as much out of the gl file as possible to make it easier to port to the other video drivers.

I think that's a good move, but huwpascoe is interested in helping with such a thing, the bounty portion of this task is finished, I think. I'll try to pin TA down on how we want to do this, either to go ahead and merge it and abstract it later or do the abstracting before merging (in which case we need to see about how best to close this issue while keeping the PR alive).

**Comment by huwpascoe** [comment]
@hizzlekizzle 
I think that's a future problem. The GL driver changes are unobtrusive as I could come up with: most of it is in one HAVE_VIDEO_LAYOUT block, and the calls into that block are also marked with the same #ifdef so anyone can extract it with a cut'n'paste if they want to.

I'd really like to get this PR closed. As happy as I am that the menu system is being cleaned up, I have to rework everything every time @twinaphex makes a commit!

**Comment by huwpascoe** [comment]
rebased. travis-ci errors are unrelated to this PR.

@twinaphex waiting on you

**Comment by twinaphex** [comment]
Hi there, sorry for the delay, we are knee deep with the new release. I hope to have that finished off by Sunday, then I hope we can start working out a solution for this so that it is scaleable across the other video drivers.

The bounty is very much appreciated and I hope that together we can get somewhere, but please give me a bit of breathing room to finish the release first. It is taking a lot of energy right now.

**Comment by huwpascoe** [comment]
Thank you for the info, the main thing is having the related issue https://github.com/libretro/RetroArch/issues/7419 closed so I can get paid. That'd take the pressure off me.

**Comment by huwpascoe** [comment]
rebased

**Comment by twinaphex** [comment]
Alright, let's give this a try. I'm still concerned how we're going to do the implementation for all these other video drivers, though.

**Comment by twinaphex** [comment]
OK, so current findings:

1) It was enabled by default in config.params.sh, given its current unstable nature on some video drivers I have changed this to be disabled by default.
2) It doesn't crash at startup with the GL driver (probably because there is an implementation), but it crashes at startup with D3D11, a driver that does not yet have an implementation for this MAME layout. I would like to request from you to fix this part at least. At no point in time should it hard crash at startup, even if an implementation of the MAME layout code is not available.

**Comment by twinaphex** [comment]
Also, you really shouldn't have enabled it yet for all the Makefile targets, especially if it doesn't even compile out of the box -

     [exec] jni/../../../../griffin/../gfx/video_layout.c:10:10: fatal error: 'verbosity.h' file not found
     [exec] #include <verbosity.h>
     [exec]          ^
     [exec] In file included from jni/../../../../griffin/griffin.c:479:
     [exec] jni/../../../../griffin/../gfx/video_layout.c:10:10: fatal error: 'verbosity.h' file not found
     [exec] #include <verbosity.h>
     [exec]          ^

It fails on Android now for instance because this is supposed to be a relative path include, you aren't supposed to be looking inside the system dir for this.

Anyway, I'll fix it up for now but it's kinda annoying this stuff wasn't properly looked at before given the pressure that was mounted to merge this.

**Comment by huwpascoe** [comment]
👍👍👍👍👍👍👍

I will test the D3D11 driver and submit a fix.

**Pull request merged**